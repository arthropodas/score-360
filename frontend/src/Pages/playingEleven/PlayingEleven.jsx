import { Box, Typography,Checkbox, Tooltip } from "@mui/material";
import React, { useEffect, useState } from "react";
import PersonIcon from "@mui/icons-material/Person";
import SubmitButton from "../../components/Common/buttons/submitButton/SubmitButton";
import { addPlayingEleven,getSquad} from "../../Services/TeamServices";
import CancelButton from "../../components/Common/buttons/cancelButton/CancelButton";
import { useLocation, useNavigate } from "react-router-dom";
import Loader from "../../components/Common/loader/Loader";
import { ToastService } from "../../components/Common/toast/ToastService";
import { AddPlayerElevenErrorMessage } from "../../utils/errorCode/ErrorCode";
const teamBoxStyle={
    width: { xs: "100%", md: "50%" },
    height: { xs: "50%", md: "100%" },
  }
const boxStyle={
    height: "100%",
    width: "auto",
    whiteSpace: "nowrap",
    overflow: "auto",
    textOverflow: "ellipsis",
    border: "2px solid #e0e0e0",
    borderRadius: "10px", // Add border radius for rounded corners
    boxShadow: "0px 4px 6px rgba(0, 0, 0, 0.1)", // Add a subtle shadow for depth
  }
  const commonBoxStyle={display: "flex", justifyContent: "center"}
const PlayingEleven = () => {
  const [playersOne, setPlayersOne] = useState([]);
  const [playersTwo, setPlayersTwo] = useState([]);
  const [selectedTeamOne, setSelectedTeamOne] = useState([]);
  const [selectedTeamTwo, setSelectedTeamTwo] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isAddButtonDisabled, setIsAddButtonDisabled] = useState(true);
  const navigate = useNavigate();
  const location = useLocation();
  let { team1, team2, matchId } = location?.state || {};
  useEffect(() => {
    const response = getSquad(team1?.id);
    response
      .then((res) => {
        setPlayersOne(res?.data);
      })
      .catch((err) => {
        console.log(err);
      });
  }, [team1?.id]);
  useEffect(() => {
    const response = getSquad(team2?.id);
    response
      .then((res) => {
        setPlayersTwo(res?.data);
      })
      .catch((err) => {
        console.log(err);
      });
  }, [team2?.id]);
  useEffect(() => {
    setIsAddButtonDisabled(
      selectedTeamOne.length === 11 && selectedTeamTwo.length === 11
        ? true
        : false
    );
  }, [selectedTeamOne, selectedTeamTwo]);

  console.log(isAddButtonDisabled);
  const handleClick = (event, email, team) => {
    if (team === "team1") {
      const selectedIndex = selectedTeamOne.indexOf(email);
      let newSelected = [];

      if (selectedIndex === -1) {
        newSelected = [...selectedTeamOne, email];
      } else {
        newSelected = selectedTeamOne.filter(
          (selectedEmail) => selectedEmail !== email
        );
      }

      setSelectedTeamOne(newSelected);
    } else if (team === "team2") {
      const selectedIndex = selectedTeamTwo.indexOf(email);
      let newSelected = [];

      if (selectedIndex === -1) {
        newSelected = [...selectedTeamTwo, email];
      } else {
        newSelected = selectedTeamTwo.filter(
          (selectedEmail) => selectedEmail !== email
        );
      }

      setSelectedTeamTwo(newSelected);
    }
  };
  const handleCancel = () => {
    navigate(-1);
  };
  console.log(selectedTeamOne, selectedTeamTwo);
  const isSelected = (email, team) => {
    if (team === "team1") {
      return selectedTeamOne.indexOf(email) !== -1;
    } else if (team === "team2") {
      return selectedTeamTwo.indexOf(email) !== -1;
    }
  };
  const data = {
    matchId: matchId,
    teamOnePlayers: selectedTeamOne,
    teamOneId: team1?.id,
    teamTwoPlayers: selectedTeamTwo,
    teamTwoId: team2?.id,
  };
  const AddElevenClick = () => {
    setIsLoading(true);
    const response = addPlayingEleven(data);
    response
      .then((res) => {
        if (res.status === 200) {
          setIsLoading(false);
          ToastService("Players added to playing eleven",'success')
          navigate(`/add-score/${matchId}`,{state:{"team1":team1,"team2":team2}})
          
        }
      })
      .catch((err) => {
setIsLoading(false);
ToastService(AddPlayerElevenErrorMessage(err.response.data.errorCode),'error')
      });
  };
  
  function renderBox(){
   return( <Box
    sx={{
      display: "flex",
      justifyContent: "center ",
      height: "100%",
      whiteSpace: "nowrap",
      overflow: "auto",
      textOverflow: "ellipsis",
      border: "2px solid #e0e0e0", // Add a solid border with a light gray color
      borderRadius: "10px", // Add border radius for rounded corners
      boxShadow: "0px 4px 6px rgba(0, 0, 0, 0.1)",
    }}
  >
    <Typography alignSelf={"center"}>
      No players to display 
    </Typography>
  </Box>)
  }
  function renderTypography(player){
    return(  <Typography
        title={player.first_name + " " + player.last_name}
        sx={{
          marginTop: 1,
          overflow: "hidden",
          textOverflow: "ellipsis",
          whiteSpace: "nowrap",
          width: { xs: "10rem", md: "10rem" },
          display: "flex",
          gap: 2,
        }}
      >
        <Box>
          <PersonIcon />
        </Box>
        <Box
          sx={{
            flexGrow: 1,
            overflow: "hidden",
            textOverflow: "ellipsis",
          }}
        >

          {player.first_name + " " + player.last_name}
        </Box>
      </Typography>)
  }
  return (
    <>
      {isLoading ? (
        <Loader />
      ) : (
        <Box
          className="container"
          sx={{ width: "55vw", height: "60vh", padding: 5, margin: "auto" }}
        >
          <Box
            sx={{...commonBoxStyle, height: "10%" }}
          >
            <Typography sx={{ fontSize: "30px", alignSelf: "center" }}>
              <b>Playing 11</b>
            </Typography>
          </Box>

          <Box
            sx={{
              height: { xs: "50%", md: "60%" },
              display: "flex",
              boxSizing: "border-box",
              gap: 5,
              flexDirection: { xs: "column", md: "row" },
            }}
          >
            <Box
              className="team1"
              sx={teamBoxStyle}
            >
              <Box sx={{ ...commonBoxStyle, mt: 2 }}>
               <Tooltip title={team1?.team_name}>
                <Typography variant="h6" sx={{fontSize:'18px'}}>
                  {team1?.team_name.length > 7
                    ? `${team1?.team_name.substring(0, 7)}.....`
                    : team1?.team_name}
                </Typography>
              </Tooltip>
              </Box>
              {playersOne.length === 0 ? (
              renderBox()
              ) : (
                <Box
                  sx={boxStyle}
                >
                  {playersOne.map((player) => (
                    <Box
                      key={player.first_name}
                      sx={{
                        display: "flex",
                        justifyContent: "flex-start",
                        flexDirection: "row",
                        marginLeft: 4,
                      }}
                    >
                      <Box sx={{ display: "flex" }}>
                        <Checkbox
                          onClick={(event) =>
                            handleClick(event, player.email, "team1")
                          }
                          checked={isSelected(player.email, "team1")}
                          data-testid="playername"
                        />
                       { renderTypography(player)}
                      </Box>
                    </Box>
                  ))}
                </Box>
              )}
            </Box>

            <Box
              className="team2"
              sx={teamBoxStyle}
            >
              <Box sx={{ display: "flex", justifyContent: "center", mt: 2 }}>
              <Tooltip title={team1?.team_name}>
                <Typography variant="h6" sx={{fontSize:'18px'}}>
                  {team2?.team_name.length > 7
                    ? `${team2?.team_name.substring(0, 7)}.....`
                    : team2?.team_name}
                </Typography>
              </Tooltip>
              </Box>
              {playersTwo.length === 0 ? (
             renderBox()
              ) : (
                <Box
                  sx={boxStyle}
                >
                  {playersTwo.map((player, index) => (
                    <Box
                      key={player.first_name}
                      sx={{
                        ...commonBoxStyle,
                        flexDirection: "row",
                        marginLeft: 4,
                      }}
                    >
                      <Box sx={{ display: "flex" }}>
                        <Checkbox
                          onClick={(event) =>
                            handleClick(event, player.email, "team2")
                          }
                          checked={isSelected(player.email, "team2")}
                          data-testid="playername"
                        />

                       {renderTypography(player)}
                      </Box>
                    </Box>
                  ))}
                </Box>
              )}
            </Box>
          </Box>
          <Box sx={ commonBoxStyle}>
            <Box
              sx={{...commonBoxStyle, gap: 2, mt: 10 }}
            >
              <CancelButton name="Cancel " onClick={handleCancel} />

              <SubmitButton name="Add" onClick={AddElevenClick} />
            </Box>
          </Box>
        </Box>
      )}
    </>
  );
};

export default PlayingEleven;
