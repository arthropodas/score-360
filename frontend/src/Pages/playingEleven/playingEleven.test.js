import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import PlayingEleven from './PlayingEleven';
import "@testing-library/jest-dom";
import { BrowserRouter } from 'react-router-dom';

// Mocking TeamServices module
jest.mock('../../Services/TeamServices', () => ({
 getSquad: jest.fn().mockResolvedValue({
   data: [
     { id: 1, name: 'Player 1' },
     { id: 2, name: 'Player 2' },
     // Add more player objects as needed
   ]
 }),
 addPlayingEleven: jest.fn().mockResolvedValue({ success: true }), // Mock addPlayingEleven function
}));


describe('PlayingEleven Component', () => {
 const mockSetElevenOpen = jest.fn();
 const mockMatchId = 'mockMatchId';
 const mockTeams = {
   team1: { id: 'team1Id' },
   team2: { id: 'team2Id' },
 };
 const mockData = {
   matchId: mockMatchId,
   teamOnePlayers: [],
   teamOneId: mockTeams?.team1?.id,
   teamTwoPlayers: [],
   teamTwoId: mockTeams?.team2?.id,
 };


 test('renders PlayingEleven component correctly', async () => {
   const { getByText } = render(
<BrowserRouter>     <PlayingEleven
       setElevenOpen={mockSetElevenOpen}
       matchId={mockMatchId}
       teams={mockTeams}
     /></BrowserRouter>
   );


  
   expect(getByText('Add')).toBeInTheDocument();
   expect(getByText('Cancel')).toBeInTheDocument();
 });


 test('selects and deselects players from team one', async () => {
   const { getAllByTestId } = render(
    <BrowserRouter>     <PlayingEleven
           setElevenOpen={mockSetElevenOpen}
           matchId={mockMatchId}
           teams={mockTeams}
         /></BrowserRouter>
       );
    // Wait for the players to be populated
 await waitFor(() => expect(getAllByTestId('playername')[0]).toBeTruthy());


 // Click on the first player checkbox
 fireEvent.click(getAllByTestId('playername')[0]);




   // Check if the player is deselected
   expect(getAllByTestId('playername')[0]).not.toBeChecked();
   expect(getAllByTestId('playername')[1]).not.toBeChecked();

   
 });


 test('handles cancel action', () => {
   const { getByText } = render(
    <BrowserRouter>     <PlayingEleven
    setElevenOpen={mockSetElevenOpen}
    matchId={mockMatchId}
    teams={mockTeams}
  /></BrowserRouter>
   );


   // Click on the cancel button
   fireEvent.click(getByText('Cancel'));


 });


 test('handles click to add eleven with error', async () => {
   const { getByText } = render(
    <BrowserRouter>     <PlayingEleven
    setElevenOpen={mockSetElevenOpen}
    matchId={mockMatchId}
    teams={mockTeams}
  /></BrowserRouter>
   );


   // Simulate clicking on the Add button
   fireEvent.click(getByText('Add'));
 });


 // Add more test cases as needed...
});


