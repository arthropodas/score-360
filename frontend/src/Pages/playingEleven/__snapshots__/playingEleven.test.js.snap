// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayingEleven Component renders PlayingEleven component correctly 1`] = `
<div>
  <div
    class="container MuiBox-root css-1yh90lj"
  >
    <div
      class="MuiBox-root css-kbk2hk"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1a9fpgg-MuiTypography-root"
      >
        <b>
          Playing 11
        </b>
      </p>
    </div>
    <div
      class="MuiBox-root css-1uemltz"
    >
      <div
        class="team1 MuiBox-root css-1c9tozj"
      >
        <div
          class="MuiBox-root css-kl1xo3"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Team 1
          </p>
        </div>
        <div
          class="MuiBox-root css-sszhsl"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-6dc9vr-MuiTypography-root"
          >
            No players to display
          </p>
        </div>
      </div>
      <div
        class="team2 MuiBox-root css-1c9tozj"
      >
        <div
          class="MuiBox-root css-kl1xo3"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Team 2
          </p>
        </div>
        <div
          class="MuiBox-root css-sszhsl"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-6dc9vr-MuiTypography-root"
          >
            No players to display
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1l4w6pd"
    >
      <div
        class="MuiBox-root css-15gfldf"
      >
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-zxijvj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Cancel 
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-8xv09s-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Add
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PlayingEleven Component renders correctly 1`] = `
<div>
  <div
    class="container MuiBox-root css-1yh90lj"
  >
    <div
      class="MuiBox-root css-kbk2hk"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1a9fpgg-MuiTypography-root"
      >
        <b>
          Playing 11
        </b>
      </p>
    </div>
    <div
      class="MuiBox-root css-1uemltz"
    >
      <div
        class="team1 MuiBox-root css-1c9tozj"
      >
        <div
          class="MuiBox-root css-kl1xo3"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Team 1
          </p>
        </div>
        <div
          class="MuiBox-root css-sszhsl"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-6dc9vr-MuiTypography-root"
          >
            No players to display
          </p>
        </div>
      </div>
      <div
        class="team2 MuiBox-root css-1c9tozj"
      >
        <div
          class="MuiBox-root css-kl1xo3"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Team 2
          </p>
        </div>
        <div
          class="MuiBox-root css-sszhsl"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-6dc9vr-MuiTypography-root"
          >
            No players to display
          </p>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-1l4w6pd"
    >
      <div
        class="MuiBox-root css-15gfldf"
      >
        <div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-zxijvj-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Cancel 
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-8xv09s-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Add
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
